### PR TITLE
Add misorientation clustering notebook

### PR DESCRIPTION
**Pull request by hakonanes**
Signed-off-by: Håkon Wiik Ånes <hwaanes@gmail.com>

#### Description of the change
Add the misorientation clustering notebook from the [orix-demos](https://github.com/pyxem/orix-demos/blob/master/03%20-%20Clustering%20Misorientations.ipynb). This is the final notebook in the demos that aren't in the user guide, and we can hence archive (and eventually delete) the orix-demos repository after this PR is merged.

The notebook is stored with cell outputs, so the code is not run by `nbsphinx` on Read the Docs, neither locally nor on their server.

#### Progress of the PR
- [n/a] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [n/a] Unit tests with pytest for all lines
- [n/a] Clean code style by [running black via pre-commit](https://orix.readthedocs.io/en/latest/contributing.html#code-style)

#### For reviewers

- [ ] The PR title is short, concise, and will make sense 1 year later.
- [ ] New functions are imported in corresponding `__init__.py`.
- [ ] New features, API changes, and deprecations are mentioned in the
      unreleased section in `CHANGELOG.rst`.

### CODE REVIEW

**Review by harripj — CHANGES_REQUESTED** [review]
Very nice notebook, a few suggestions below. One more suggestion, once the data is loaded I think it would be nice to show an IPF map of the region, that way the user can compare to the final grain boundary map. What do you think?

File: doc/clustering_misorientations.ipynb [hunk 1]
```diff
@@ -0,0 +1,754 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#mismisori = (~mori2).outer(mori2)\n",
+    "#mismisori_equiv = D6.outer(~mori2).outer(D6).outer(D6).outer(mori2).outer(D6)\n",
```
[the remark this answers is not in the document]
Comment (harripj) [reply doc/clustering_misorientations.ipynb:246]: Yes something like this. I'm not suggesting to change the code, just curious!

[the remark this answers is not in the document]
Comment (hakonanes) [reply doc/clustering_misorientations.ipynb:246]: Good suggestion, it reduces the number of computations, which is always good! Will do the change.

[the remark this answers is not in the document]
Comment (harripj) [reply doc/clustering_misorientations.ipynb:246]: For example, for the main misorientation distance calculation, the following produces the same output as calculated (code taken from the notebook):
```python
>>> mismisori = (~mori).outer(mori)

>>> # NB the 2 lines below are the memory intense operation, so left commented out
>>> # mismisori_equiv = D6.outer(~mori2).outer(D6).outer(D6).outer(mori2).outer(D6)
>>> # D = mismisori_equiv.angle.data.min(axis=(0, 2, 3, 5))

>>> # load the precomputed file instead
>>> D = np.load("misori-distance((100, 200)).npy")

>>> # alternative with 2 less dot products
>>> mismisori_equiv1 = D6.outer(~mori).outer(D6.outer(mori))
>>> D1 = mismisori_equiv1.angle.min(axis=(0, 2))

>>> np.allclose(D, D1)
True
```

It strikes me that this is possible because in the 4 dot product equation the preceding and succeeding symmetries for each misorientation are the same in this case, ie. `D6`.

[the remark this answers is not in the document]
Comment (hakonanes) [reply doc/clustering_misorientations.ipynb:246]: Yes, and I think the authors (@pc494) wanted to show the general approach and not assume identical symmetry. Although fewer dot products are computed, this almost fills my 16 GB RAM, so we can't ask Read the Docs server to do it. But we could add it in the commented out section.

[the remark this answers is not in the document]
Comment (harripj) [reply doc/clustering_misorientations.ipynb:246]: Thanks for clarifying! I don't think the notebook needs changing, just curious.

File: doc/clustering_misorientations.ipynb [hunk 2]
```diff
@@ -0,0 +1,754 @@
...
+    "from sklearn.cluster import DBSCAN\n",
+    "\n",
+    "# Colorisation and visualisation\n",
+    "from skimage.color import label2rgb\n",
+    "from matplotlib.colors import to_rgb\n",
+    "from matplotlib.lines import Line2D\n",
+    "\n",
+    "# Import orix classes\n",
+    "from orix import plot\n",
+    "from orix.quaternion import Orientation, OrientationRegion, Misorientation, Rotation\n",
```
[the remark this answers is not in the document]
Comment (hakonanes) [reply doc/clustering_misorientations.ipynb]: Thanks, removed the two first, since `Vector3d` is now used to specify sample directions when colouring orientations.

File: doc/clustering_misorientations.ipynb [hunk 3]
```diff
@@ -0,0 +1,754 @@
...
+    "Load orientation mapping data specified in a standard CTF file as euler angles in degrees, following the Bunge convention"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "filepath = \"/home/hakon/kode/orix_test/Ti_orientations.ctf\"\n",
```
[the remark this answers is not in the document]
Comment (hakonanes) [reply doc/clustering_misorientations.ipynb]: Right, these are now shortened to just the file names. I intended to leave them in because it will be simpler for me to run them in the future to check functionality.

File: doc/clustering_misorientations.ipynb [hunk 4]
```diff
@@ -0,0 +1,754 @@
...
+   "outputs": [],
+   "source": [
+    "mori_all = Misorientation(~ori[:, :-1] * ori[:, 1:])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Keep only misorientations assumed to be on boundaries, i.e. with a\n",
```
[the remark this answers is not in the document]
Comment (hakonanes) [reply doc/clustering_misorientations.ipynb]: Good point, have rephrased slightly to make this clear.

File: doc/clustering_misorientations.ipynb [hunk 5]
```diff
@@ -0,0 +1,754 @@
...
+   "outputs": [],
+   "source": [
+    "ori = ori.reshape(381, 507)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Select a subset of the orientations to a suitable size for this demonstration"
```
[the remark this answers is not in the document]
Comment (hakonanes) [reply doc/clustering_misorientations.ipynb]: I agree.

File: doc/clustering_misorientations.ipynb [hunk 6]
```diff
@@ -0,0 +1,754 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#from itertools import combinations_with_replacement as icombinations\n",
+    "#from tqdm import tqdm_notebook\n",
+    "#D = np.empty((misori.size, misori.size))\n",
```
[the remark this answers is not in the document]
Comment (hakonanes) [reply doc/clustering_misorientations.ipynb]: Thanks, this should be correct now.

File: doc/clustering_misorientations.ipynb [hunk 7]
```diff
@@ -0,0 +1,754 @@
...
+    "#\n",
+    "#for i, j in tqdm_notebook(list(icombinations(range(mori2.size), 2))):\n",
+    "#    m1, m2 = mori2[i], mori2[j]\n",
+    "#    mismisori = D6.outer(~m1).outer(D6).outer(D6).outer(m2).outer(D6)\n",
+    "#    d = mismisori.angle.data.min(axis=(0, 2, 3, 5))\n",
```
[the remark this answers is not in the document]
Comment (hakonanes) [reply doc/clustering_misorientations.ipynb]: Thanks, this should be correct now.

File: doc/clustering_misorientations.ipynb [hunk 8]
```diff
@@ -0,0 +1,754 @@
...
+    "### Load a distance matrix we computed for you"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "filepath_2 = \"/home/hakon/kode/orix_test/misori-distance((100, 200)).npy\"\n",
```
[the remark this answers is not in the document]
Comment (hakonanes) [reply doc/clustering_misorientations.ipynb]: Done.

File: doc/clustering_misorientations.ipynb (hunk 1 above)
[the remark this answers is not in the document]
Comment (hakonanes) [reply doc/clustering_misorientations.ipynb]: I added this approach as a second option, specifying that this only works when the misorientation symmetries are the same.

File: doc/clustering_misorientations.ipynb [hunk 9]
```diff
@@ -0,0 +1,754 @@
...
+    "Compute misorientations"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "mori_all = Misorientation(~ori[:, :-1] * ori[:, 1:])"
```
[the remark this answers is not in the document]
Comment (harripj) [reply doc/clustering_misorientations.ipynb:220]: > Tools like finding indices of nearest neighbours in a 2D map would be useful here. I've been thinking to add such a function(s), to be used for KAM calculations etc., but never found the time for it.

I had a go at something like this before, it may be useful to you:
```python
from orix.quaternion import Orientation
import numpy as np
order = 1
o = Orientation.random((200, 100))
from skimage.morphology import disk
slices = tuple(slice(order, s-order) for s in o.shape)
indices = np.stack(np.mgrid[slices], axis=-1)

footprint = disk(order)
footprint_indices = np.column_stack(np.nonzero(footprint)) - np.array(footprint.shape) // 2

indices_all = indices[..., np.newaxis, :] + footprint_indices
m = o[tuple(indices_all.T)].transpose(2, 1, 0) - Orientation(o[slices].data[..., np.newaxis, :])
```
I didn't finish it off as it requires the mean to be taken along a certain axis (which is currently not implemented). But it's a start if you're looking to do something similar.

File: doc/clustering_misorientations.ipynb (hunk 7 above)
[the remark this answers is not in the document]
Comment (harripj) [reply doc/clustering_misorientations.ipynb]: I think there is still a reference to `mori2` in the loop in the latest build :).

[the remark this answers is not in the document]
Comment (hakonanes) [reply doc/clustering_misorientations.ipynb]: Ah, sorry. Thank you for looking closely!

File: doc/clustering_misorientations.ipynb (hunk 9 above)
[the remark this answers is not in the document]
Comment (hakonanes) [reply doc/clustering_misorientations.ipynb:220]: Thanks, I'll keep this comment in mind if I find the time!

Adding an `axis` argument to various methods, like `mean`, `unique` etc., will be very powerful. Something for the future...

**Review by harripj — APPROVED** [review]

File: doc/clustering_misorientations.ipynb (hunk 1 above)
[the remark this answers is not in the document]
Comment (harripj) [reply doc/clustering_misorientations.ipynb]: I will leave this open if @pc494 would like to add any comments!

[the remark this answers is not in the document]
Comment (pc494) [reply doc/clustering_misorientations.ipynb]: Happy with this.

**Review by pc494 — APPROVED** [review]
This is a really nice tidy up, merging
